controllers/watcher_handler.go

```diff
@@ -0,0 +1,96 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2023 Datadog, Inc.
+
+package controllers
+
+import (
+	"fmt"
+
+	chaosv1beta1 "github.com/DataDog/chaos-controller/api/v1beta1"
+	"github.com/DataDog/chaos-controller/watchers"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type newWatcher func(name string, r *DisruptionReconciler, instance *chaosv1beta1.Disruption) (watchers.Watcher, error)
+
+var watcherList = map[string]newWatcher{
+	"chaos-pod": func(name string, r *DisruptionReconciler, instance *chaosv1beta1.Disruption) (watchers.Watcher, error) {
+		return watchers.NewChaosPodWatcher(name, r.Recorder, instance, r.log)
+	},
+	"disruption-target": func(name string, r *DisruptionReconciler, instance *chaosv1beta1.Disruption) (watchers.Watcher, error) {
+		return watchers.NewDisruptionTargetWatcher(name, r.Recorder, r.Reader, r.EnableObserver, instance, r.MetricsSink, r.log)
+	},
+}
+
+// manageInstanceWatchers manage the watchers for each disruption instance
+func (r *DisruptionReconciler) manageInstanceWatchers(instance *chaosv1beta1.Disruption) error {
+	r.log.Debugf("start to manage watchers for the %s disruption", instance.Name)
+
+	if len(watcherList) == 0 {
+		return nil
+	}
+
+	if err := r.WatcherManager.RemoveExpiredWatchers(); err != nil {
```

Comment:
For me it's an issue to have this on the critical path of creating a new cache handler.
If this returns an error, you can't create any cache anymore.

This feels counter intuitive.

Let's see how it will look like with the suggested factory change and we will see.

Reply:
After investigation, the method `RemoveExpiredWatchers` of the `WatcherManager` does not return an error. So I have removed the return error and now it is not blocker anymore.